ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,152 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
-        nodesToExclude.add(nodeWithMostCapacity);
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}.  Does not retry if fewer than
+   * {@code replicaCountPerDatacenter} disks cannot be allocated.
+   *
+   * @param replicaCountPerDatacenter how many replicas to attempt to allocate in the datacenter
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param disableRackAwareness if {@code true}, do not attempt a rack-aware allocation
+   * @return A list of {@link Disk}s
+   */
+  private List<Disk> allocateDisksForPartition(int replicaCountPerDatacenter, long replicaCapacityInBytes,
+      Datacenter datacenter, boolean disableRackAwareness) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
+    if (!datacenter.isRackAware() && !disableRackAwareness) {
+      throw new IllegalArgumentException("disableRackAwareness is false, but the datacenter: " + datacenter.getName()
+          + " does not have rack information");
+    }
+    Set<DataNode> nodesToExclude = new HashSet<>();
+    List<DataNode> dataNodes = new ArrayList<>(datacenter.getDataNodes());
+    for (int i = 0; i < replicaCountPerDatacenter; i++) {
+      List<Disk> diskCandidates =
+          getRandomDiskCandidateSample(dataNodes, nodesToExclude, replicaCapacityInBytes, disableRackAwareness,
+              NUM_CHOICES);
+      if (diskCandidates.size() > 0) {
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
+        disksToAllocate.add(diskToAdd);
+        nodesToExclude.add(diskToAdd.getDataNode());
+      } else {
+        break;
       }
     }
-    return allocatedDisks;
+    return disksToAllocate;
   }
 
-  // Best effort (or less) allocation of partitions. I.e., size of returned list may be less than numPartitions.
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}. Retry a non rack-aware allocation
+   * in certain cases described below if {@code bestEffort} is enabled.
+   *
+   * @param replicaCountPerDatacenter how many replicas to attempt to allocate in the datacenter
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param disableRackAwareness if {@code true}, do not attempt a rack-aware allocation
+   * @param bestEffort If {@code bestEffort} is {@code true} and {@code disableRackAwareness} is {@code false},
+   *                   this method will attempt a non rack-aware allocation if it cannot find a rack-aware allocation.
+   * @return a list of {@code replicaCountPerDatacenter} or fewer disks that can be allocated for a new partition in
+   *         the specified datacenter
+   */
+  private List<Disk> allocateDisksForPartitionWithRetry(int replicaCountPerDatacenter, long replicaCapacityInBytes,
```

Comment:
Taking a second look, it seems to me that we need a rethink of these arguments. What is the point of having both `bestEffort` and `disableRackAwareness`?

What if we named `bestEffort` as `retryIfNotRackAware` (or similar) and had only that argument? I think it would make everything simpler. 

For `allocatePartitions()`, if `retryIfNotRackAware` is `false`, we bail out immediately if any of the data centers returns `isRackAware()` as `false` (because it is clearly a lost cause). 
If `retryIfNotRackAware` is `true`, then for the the ones that return `false` for `isRackAware()` we simply disable rack-awareness on the first call to `allocateDisksForPartition()`. For the ones that return `true`, we use the logic in this function.